cmd/http-trace.go

```diff
@@ -0,0 +1,219 @@
+	"strings"
+	"time"
+)
+
+// recordRequest - records the first recLen bytes
+// of a given io.Reader
+type recordRequest struct {
+	// Data source to record
+	io.Reader
+	// Response body should be logged
+	logBody bool
+	// Internal recording buffer
+	buf bytes.Buffer
+}
+
+func (r *recordRequest) Read(p []byte) (n int, err error) {
+	n, err = r.Reader.Read(p)
+	if r.logBody {
+		r.buf.Write(p[:n])
+	}
+	if err != nil {
+		return n, err
+	}
+	return n, err
+}
+
+// Return the bytes that were recorded.
+func (r *recordRequest) Data() []byte {
+	return r.buf.Bytes()
+}
+
+// recordResponseWriter - records the first recLen bytes
+// of a given http.ResponseWriter
+type recordResponseWriter struct {
+	// Data source to record
+	http.ResponseWriter
+	// Response body should be logged
+	logBody bool
+	// Internal recording buffer
+	headers bytes.Buffer
+	body    bytes.Buffer
+	// The status code of the current HTTP request
+	statusCode int
+	// Indicate if headers are written in the log
+	headersLogged bool
+}
+
+// Write the headers into the given buffer
+func writeHeaders(w io.Writer, statusCode int, headers http.Header) {
+	fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
+	for k, v := range headers {
+		fmt.Fprintf(w, "%s: %s\n", k, v[0])
+	}
+}
+
+// Record the headers.
+func (r *recordResponseWriter) WriteHeader(i int) {
+	r.statusCode = i
+	if !r.headersLogged {
+		writeHeaders(&r.headers, i, r.ResponseWriter.Header())
+		r.headersLogged = true
+	}
+	r.ResponseWriter.WriteHeader(i)
+}
+
+func (r *recordResponseWriter) Write(p []byte) (n int, err error) {
+	n, err = r.ResponseWriter.Write(p)
+	// We log after calling ResponseWriter.Write() because this latter
+	// proactively prepares headers when WriteHeader is not called yet.
+	if !r.headersLogged {
+		writeHeaders(&r.headers, http.StatusOK, r.ResponseWriter.Header())
+		r.headersLogged = true
+	}
+	if r.statusCode != http.StatusOK && r.statusCode != http.StatusNoContent && r.statusCode != 0 {
+		// We always log error responses.
+		r.body.Write(p)
+		return
+	}
+	if r.logBody {
+		r.body.Write(p)
+		return
+	}
+	return n, err
+}
+
+// Calls the underlying Flush.
+func (r *recordResponseWriter) Flush() {
+	r.ResponseWriter.(http.Flusher).Flush()
+}
+
+// Return response headers.
+func (r *recordResponseWriter) Headers() []byte {
+	return r.headers.Bytes()
+}
+
+// Return response body.
+func (r *recordResponseWriter) Body() []byte {
+	return r.body.Bytes()
+}
+
+// Log headers and body.
+func httpTracelogAll(f http.HandlerFunc) http.HandlerFunc {
+	if !globalHTTPTrace {
+		return f
+	}
+	return httpTracelog(f, true)
+}
+
+// Log only the headers.
+func httpTracelogHeaders(f http.HandlerFunc) http.HandlerFunc {
+	if !globalHTTPTrace {
+		return f
+	}
+	return httpTracelog(f, false)
+}
+
+// Logs request/response headers and body.
+// Body is not logged for PutObject, PutObjectPart and GetObject.
+func httpTracelog(f http.HandlerFunc, logBody bool) http.HandlerFunc {
```

Comment:
We should call this as handler since it returns a HandlerFunc .. like `httpTraceHandler` etc.